README.md
# IPK Project 2 - ZETA: Network sniffer

## Necessary theory

The following theory is necessary to understand the project:

- Ethernet frame
- IP packet
- TCP segment
- UDP datagram
- ICMPv4 packet
- ICMPv6 packet
- ARP packet
- NDP packet
- IGMP packet
- MLD packet
- Tree data structure
- C++20 standard

## Project structure

All source files are located under `src` directory.

All required libraries are located under `lib` directory.

All tests are located under `tests` directory.

All additional documentation files are located under `doc` directory.

## Sniffer flow

Application starts by parsing command line arguments. Arguments are parsed using
`argparse` library. After parsing and validating arguments, the application
initializes the capturing using `pcap` library. Then the application enters
the main loop where it waits for packets. When a packet is captured, the
application processes it and prints it to the standard output. After the
specified number of packets is captured, the application exits.

### Packet filtering

![filter-tree-uml.png](docs/filter-tree-uml.png)

The application uses `pcap_compile` and `pcap_setfilter` functions to filter
packets. The filter is constructed from the command line arguments.

For parsing filters from command line arguments, the application uses
`FilterTree` class. The class is a tree of `FilterTree` nodes. Each node
represents a filter or a group of filters connected by logical operators.
The class provides a method `generate_filter` that generates a string
representation of the filter using tree traversal. The string is then passed
to `pcap_compile` and `pcap_setfilter` functions.

## Building

### Requirements

- pcap library
- C++20 compiler
- make

### Building

```
make
```

### Cleaning

```
make clean
```

## Usage

```
Usage: IPK Sniffer [--help] [--version] [--interface interface] [--tcp] [--udp] [--port port] [--icmp4] [--icmp6] [--arp] [--ndp] [--igmp] [--mld] [-n num]

Optional arguments:
  -h, --help                    shows help message and exits 
  -v, --version                 prints version information and exits 
  -i, --interface interface     Interface to sniff on [default: "-"]
  -t, --tcp                     TCP packets 
  -u, --udp                     UDP packets 
  -p, --port port               Port to sniff on [default: 0]
  --icmp4                       Filter by port 
  --icmp6                       Display only ICMPv6 echo request/response 
  --arp                         Display only ARP frames 
  --ndp                         Display only ICMPv6 NDP packets 
  --igmp                        Display only IGMP packets 
  --mld                         Display only MLD packets 
  -n num                        Number of packets to display [default: 1]

Author: Nikita Moiseev <xmoise01@stud.fit.vutbr.cz>
```

## Testing

### Network topology

![topology.png](docs/topology.png)

Network consists of 1 router and 2 laptops. Each laptop is connected to the
router via Wi-Fi. Router is connected to the internet via Ethernet.

### Testing environment

Laptop 1:

- OS: MacOS Ventura 13.0.1 (22A400)
- IP: 192.168.1.17
- MAC: b0:be:83:4a:c4:b0

Laptop 2:

- OS: Ubuntu 18.04.3 LTS
- IP: 192.168.1.49
- MAC: 94:08:53:2e:50:cb

Router:

- Name: O2 SmartBox

### Testing procedure

The following protocols have been tested:

- TCP
- UDP
- ICMPv4
- ICMPv6
- ARP

The following protocols have been not tested:

- NDP
- IGMP
- MLD

All tests are located under `tests` directory. Each subdirectory is a protocol
that has been tested (except `extra` directory). Each protocol directory has the
following structure:

- `[ii]-[test_name].sh` - script that runs the test
- `[ii]-[test_name].out` - actual output of the test
- `[ii]-[test_name].[n].jpg` - screenshot from wireshark with the packet
  number `n`

Protocols NDP, IGMP and MLD have not been tested because their capture did not
succeed after several attempts each of 10 minutes.

Each test case was run in the command line manually and the output was saved to
the file. Then the output was compared to the packets captured by wireshark.
The packet information such as source and destination IP addresses, source and
destination MAC addresses and ports were compared. The data in the packets was
also compared.

Also `tests` folder contains `extra` directory. This directory contains
additional tests: testing the application with invalid arguments, testing
capture on multiple protocols and error handling.

### Example of tests

#### Test case: `tests/tcp/01-port-specified-1-packet`

Run the test:

```sh
./ipk-sniffer -i en0 --tcp -p 22
```

Actual output:

```
timestamp: 2023-04-17T19:50:36.270+02:00
src MAC: b0:be:83:4a:c4:b0
dst MAC: 98:77:e7:50:60:5e
frame length: 134 bytes
src IP: 192.168.1.17
dst IP: 147.229.176.14
src port: 64168
dst port: 22

0x0000: 98 77 e7 50 60 5e b0 be 83 4a c4 b0 08 00 45 00  .w.P`^...J....E.
0x0010: 00 78 00 00 40 00 40 06 34 d3 c0 a8 01 11 93 e5  .x..@.@.4.......
0x0020: b0 0e fa a8 00 16 71 3c 60 c4 c8 0c 4d 8e 80 18  ......q<`...M...
0x0030: 08 00 95 5a 00 00 01 01 08 0a ab 89 dd ea c8 1e  ...Z............
0x0040: aa 97 e4 1b 57 a6 a9 3f 26 45 ce 1b ac 94 fb 8a  ....W..?&E......
0x0050: f6 3a c5 de 08 f1 c2 16 c5 cb bd 2e 1a 4b c3 df  .:...........K..
0x0060: c8 3d f7 b4 80 4f cf 1e 6b b0 99 a1 ea cd c2 1c  .=...O..k.......
0x0070: e7 28 db 9f b2 b0 63 fe e9 65 97 8c ea f5 43 9a  .(....c..e....C.
0x0080: b1 44 fc 1e 93 1e                                .D....
```

WireShark screenshot with the packet:

![01-port-specified-1-packet.png](tests/tcp/01-port-specified-1-packet.jpg)

#### Test case: `tests/arp/01-1-packet`

Run the test:

```sh
./ipk-sniffer -i en0 --arp
```

Actual output:

```
timestamp: 2023-04-17T14:24:29.076+02:00
src MAC: 98:77:e7:50:60:5e
dst MAC: b0:be:83:4a:c4:b0
frame length: 52 bytes
src IP: 192.168.1.138
dst IP: 192.168.1.17

0x0000: b0 be 83 4a c4 b0 98 77 e7 50 60 5e 08 06 00 01  ...J...w.P`^....
0x0010: 08 00 06 04 00 01 98 77 e7 50 60 5e c0 a8 01 8a  .......w.P`^....
0x0020: 00 00 00 00 00 00 c0 a8 01 11 00 00 00 00 00 00  ................
0x0030: 00 00 00 00                                      ....
```

WireShark screenshot with the packet:

![01-arp.png](tests/arp/01-1-packet.jpg)

#### Test case: `tests/icmp4/01-1-packet`

Run the test:

```sh
./ipk-sniffer -i en0 --icmp4
```

Actual output:

```
timestamp: 2023-04-17T20:02:49.254+02:00
src MAC: b0:be:83:4a:c4:b0
dst MAC: 94:08:53:2e:50:cb
frame length: 98 bytes
src IP: 192.168.1.17
dst IP: 192.168.1.49

0x0000: 94 08 53 2e 50 cb b0 be 83 4a c4 b0 08 00 45 00  ..S.P....J....E.
0x0010: 00 54 14 4c 00 00 40 01 e2 ca c0 a8 01 11 c0 a8  .T.L..@.........
0x0020: 01 31 08 00 17 a9 26 ce 00 01 64 3d 89 c9 00 03  .1....&...d=....
0x0030: e0 7a 08 09 0a 0b 0c 0d 0e 0f 10 11 12 13 14 15  .z..............
0x0040: 16 17 18 19 1a 1b 1c 1d 1e 1f 20 21 22 23 24 25  .......... !"#$%
0x0050: 26 27 28 29 2a 2b 2c 2d 2e 2f 30 31 32 33 34 35  &'()*+,-./012345
0x0060: 36 37                                            67
```

WireShark screenshot with the packet:

![01-icmp4.png](tests/icmp4/01-1-packet.jpg)

## Bibliography

- An introduction to libpcap: https://www.tcpdump.org/pcap.html
- libpcap documentation: https://www.tcpdump.org/manpages/pcap.3pcap.html
- pcap-filter man page: https://www.tcpdump.org/manpages/pcap-filter.7.html
- Argument Parser for Modern C++: http://github.com/p-ranav/argparse
- RFC 3339 - Date and Time on the Internet: Timestamps
- IPK lectures: https://moodle.vut.cz/mod/folder/view.php?id=289124

CHANGELOG.md
# Change Log

## Arguments parsing module

All required arguments and their proper parsing is supported by sniffer.

* Add `argparse` library for parsing command line arguments
* Add `src/arguments.cpp` and `src/arguments.h` files
* Add `--help` argument
* Add `--version` argument
* Add `-i / --interface` argument
* Add `-n` argument
* Add `-t / --tcp` argument
* Add `-u / --udp` argument
* Add `-p` argument
* Add `--icmp4` argument
* Add `--icmp6` argument
* Add `--arp` argument
* Add `--ndp` argument
* Add `--igmp` argument
* Add `--mld` argument

## PCAP interfaces module

All required device validation and information is supported by sniffer.

* Add `pcap` library for capturing packets
* Add `src/device.cpp` and `src/device.h` files
* Add `NoInterfaceException` class for handling no interface selection error
* Add `get_devices` function for getting all available devices
* Add `print_device` function for printing device name
* Add `print_interfaces` function for printing all available devices
* Add `is_device_valid` for checking if device exists

## Packet filtering module

All required filtering options are supported by sniffer.

* Add `src/filter_tree.cpp` and `src/filter_tree.h` files
* Add `FILTER_TREE_TYPE` enum for representing filter tree node type
* Add `FilterTree` class for representing filter tree
* Add `FilterTree::FilterTree` constructor for creating filter tree node with
  the following options:
    * `FilterTree(FILTER_TREE_TYPE type, std::string value)` create node with
      type `type` and value `value`
    * `FilterTree(FILTER_TREE_TYPE type)` create node with type `type` and
      empty value
    * `FilterTree(std::string value)` create node with
      type `FILTER_TREE_TYPE::FILTER` and value `value`
* Add private `FilterTree::type` member for storing filter tree node type
* Add private `FilterTree::filter` member for storing filter tree node value
* Add private `FilterTree::children` member for storing filter tree node
  children
* Add private `FilterTree::process_current_node` method for processing current
  node
* Add `FilterTree::~FilterTree` destructor for deleting filter tree node
* Add `FilterTree::add_child` method for adding child node
* Add `FilterTree::get_children` method for getting children nodes
* Add `FilterTree::generate_filter` method for generating filter string

## Packet capturing module

Required output format and packet processing is supported by sniffer.

* Add `src/main.cpp` file
* Add `main` function for starting the application
* Add `generate_filter` function for generating filter string from command line
  arguments
* Add `sniff` function for capturing packets
* Add `packet_handler` function for processing captured packets
* Add `print_payload` function for printing packet payload
* Add `format_timestamp` function for formatting timestamp to RFC 3339 format
* Add `format_ip_addresses` function for formatting IP addresses
* Add `format_mac` function for formatting MAC addresses
* Add `format_ports` function for formatting ports

tests/arp/01-1-packet.sh
./ipk-sniffer -i en0 --arp

tests/icmp4/01-1-packet.sh
./ipk-sniffer -i en0 --icmp4

tests/icmp6/01-1-packet.sh
./ipk-sniffer -i en0 --icmp6

src/devices.h
/**
 * Devices module
 * @file: devices.h
 * @date: 16.04.2023
 */

#ifndef IPK_PROJECT_2_DEVICES_H
#define IPK_PROJECT_2_DEVICES_H

#include <vector>
#include <string>
#include <pcap.h>

/**
 * Exception thrown when interface is not found
 */
class NoInterfaceException : public std::exception {
private:
    const char *interface_name;

public:
    /**
     * Constructor
     * @param interface_name name of the interface
     */
    explicit NoInterfaceException(const char *interface_name) : interface_name(interface_name) {}

    /**
     * Get exception message
     * @return exception message
     */
    [[nodiscard]] const char *what() const noexcept override;
};

/**
 * Get all available devices
 * @return list of devices
 */
pcap_if_t *get_devices();

/**
 * Print device name
 * @param device device to print
 */
void print_device(pcap_if_t *device);

/**
 * Print all available devices
 */
void print_interfaces();

/**
 * Check if device is valid
 * @param device_name device name
 * @return true if device is valid, false otherwise
 */
bool is_device_valid(const std::string &device_name);

#endif// IPK_PROJECT_2_DEVICES_H


src/devices.cpp
/**
 * Devices module
 * @file: devices.cpp
 * @date: 16.04.2023
 */

#include <iostream>
#include "devices.h"

const char *NoInterfaceException::what() const noexcept {
    auto *message = new std::string();

    message->append("No interface with name: ");
    message->append(interface_name);

    return (char *) message->c_str();
}

pcap_if_t *get_devices() {
    pcap_if_t *all_devices;
    char err_buf[PCAP_ERRBUF_SIZE];

    if (pcap_findalldevs(&all_devices, err_buf) == -1) {
        throw std::runtime_error("Error in pcap_findalldevs: " + std::string(err_buf));
    }

    return all_devices;
}

void print_device(pcap_if_t *device) { std::cout << "Name: " << device->name << std::endl; }

void print_interfaces() {
    pcap_if_t *devices = get_devices();

    std::cout << "Available interfaces:" << std::endl;

    for (pcap_if_t *device = devices; device; device = device->next) { print_device(device); }
}

bool is_device_valid(const std::string &device_name) {
    pcap_if_t *devices = get_devices();

    for (pcap_if_t *device = devices; device; device = device->next) {
        if (device_name == device->name) { return true; }
    }

    throw NoInterfaceException(device_name.c_str());
}


src/arguments.cpp
/**
 * Arguments processing using argparse library http://github.com/p-ranav/argparse
 * @file: arguments.cpp
 * @date: 16.04.2023
 */

#include "arguments.h"

argparse::ArgumentParser *process_arguments(int argc, char **argv) {
    auto *program = new argparse::ArgumentParser("IPK Sniffer");

    program->add_epilog("Author: Nikita Moiseev <xmoise01@stud.fit.vutbr.cz>");

    program->add_argument("-i", "--interface").help("Interface to sniff on").default_value("-").metavar("interface");

    program->add_argument("-t", "--tcp").help("TCP packets").default_value(false).implicit_value(true);
    program->add_argument("-u", "--udp").help("UDP packets").default_value(false).implicit_value(true);

    program->add_argument("-p", "--port")
            .help("Port to sniff on")
            .default_value(0)
            .metavar("port")
            .action([](const std::string &value) {
                try {
                    int port = std::stoi(value);

                    if (port < 0 || port > 65535) { throw std::runtime_error("Port must be in range 0-65535"); }

                    return port;
                } catch (std::exception &e) { throw std::runtime_error("Cannot resolve port"); } catch (...) {
                    throw std::runtime_error("Cannot resolve port");
                }
            });

    program->add_argument("--icmp4").help("Filter by port").default_value(false).implicit_value(true);
    program->add_argument("--icmp6")
            .help("Display only ICMPv6 echo request/response")
            .default_value(false)
            .implicit_value(true);
    program->add_argument("--arp").help("Display only ARP frames").default_value(false).implicit_value(true);
    program->add_argument("--ndp").help("Display only ICMPv6 NDP packets").default_value(false).implicit_value(true);
    program->add_argument("--igmp").help("Display only IGMP packets").default_value(false).implicit_value(true);
    program->add_argument("--mld").help("Display only MLD packets").default_value(false).implicit_value(true);

    program->add_argument("-n")
            .help("Number of packets to display")
            .default_value(1)
            .metavar("num")
            .action([](const std::string &value) {
                int num = std::stoi(value);
                return num;
            });

    try {
        program->parse_args(argc, argv);
    } catch (const std::runtime_error &err) {
        std::cout << err.what() << std::endl;
        exit(0);
    }

    return program;
}


src/filter_tree.h
/**
 * Filter tree class
 * @file: filter_tree.h
 * @date: 16.04.2023
 */

#ifndef IPK_PROJECT_2_FILTER_TREE_H
#define IPK_PROJECT_2_FILTER_TREE_H

#include <vector>
#include <string>

/**
 * Filter tree type enum
 */
enum FILTER_TREE_TYPE { FILTER_TREE_TYPE_AND, FILTER_TREE_TYPE_OR, FILTER_TREE_TYPE_FILTER };

/**
 * Filter tree class
 */
class FilterTree {
private:
    /** Filter tree type */
    FILTER_TREE_TYPE type;
    /** Filter string */
    std::string *filter = nullptr;
    /** Children (in case of AND or OR) */
    std::vector<FilterTree *> *children = nullptr;

    /**
     * Process current node
     * @return filter string
     */
    std::string process_current_node();

public:
    /**
     * Constructor with type without filter string
     * @param type filter tree type
     */
    explicit FilterTree(FILTER_TREE_TYPE type);

    /**
     * Constructor with filter string and default type FILTER_TREE_TYPE_FILTER
     * @param filter filter string
     */
    explicit FilterTree(std::string *filter);

    /**
     * Constructor with type and filter string
     * @param type filter tree type
     * @param filter filter string
     */
    FilterTree(FILTER_TREE_TYPE type, std::string *filter);

    /**
     * Destructor
     */
    ~FilterTree();

    /**
     * Add child to the tree
     * @param child child to add
     */
    void add_child(FilterTree *child);

    /**
     * Get children
     * @return children
     */
    [[nodiscard]] std::vector<FilterTree *> *get_children() const;

    /**
     * Generate filter string
     * @return filter string
     */
    std::string generate_filter();
};

#endif// IPK_PROJECT_2_FILTER_TREE_H


src/filter_tree.cpp
/**
 * Filter tree class
 * @file: filter_tree.cpp
 * @date: 16.04.2023
 */

#include "filter_tree.h"

FilterTree::FilterTree(FILTER_TREE_TYPE type) : type(type) {}

FilterTree::FilterTree(std::string *filter) : type(FILTER_TREE_TYPE_FILTER), filter(filter) {}

FilterTree::FilterTree(FILTER_TREE_TYPE type, std::string *filter) : type(type), filter(filter) {}

FilterTree::~FilterTree() {
    delete children;
    delete filter;
}

void FilterTree::add_child(FilterTree *child) {
    if (children == nullptr) { children = new std::vector<FilterTree *>(); }

    children->push_back(child);
}

std::vector<FilterTree *> *FilterTree::get_children() const { return children; }

std::string FilterTree::generate_filter() {
    if (children != nullptr && !children->empty())
        for (auto *child: *children) { child->generate_filter(); }

    return process_current_node();
}

std::string FilterTree::process_current_node() {
    if (type == FILTER_TREE_TYPE_FILTER) { return *filter; }

    if (type == FILTER_TREE_TYPE_AND) {
        std::string result = "(";

        if (children == nullptr) return "";

        for (auto *child: *children) {
            result.append(child->generate_filter());
            result.append(" and ");
        }

        result.pop_back();
        result.pop_back();
        result.pop_back();
        result.pop_back();
        result.pop_back();
        result.append(")");

        return result;
    }

    if (type == FILTER_TREE_TYPE_OR) {
        std::string result = "(";

        if (children == nullptr) return "";

        for (auto *child: *children) {
            result.append(child->generate_filter());
            result.append(" or ");
        }

        result.pop_back();
        result.pop_back();
        result.pop_back();
        result.pop_back();
        result.append(")");

        return result;
    }

    return "";
}


src/main.cpp
/**
 * IPK project 2 - sniffer
 * Author: Nikita Moiseev <xmoise01> <xmoise01@stud.fit.vutbr.cz>
 */

#include <pcap.h>
#include <chrono>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <net/ethernet.h>
#include <netinet/ip6.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/socket.h>
#include <time.h>
#include "arguments.h"
#include "devices.h"
#include "filter_tree.h"

/**
 * Generate filter string from arguments
 * @param arguments arguments
 * @return filter string
 */
std::string generate_filter(argparse::ArgumentParser *arguments) {
    auto *filterTree = new FilterTree(FILTER_TREE_TYPE_OR);

    int port = arguments->get<int>("--port");
    bool has_tcp = arguments->get<bool>("--tcp");
    bool has_udp = arguments->get<bool>("--udp");

    if (port != 0) {
        auto *portFilterTree = new FilterTree(FILTER_TREE_TYPE_AND);

        portFilterTree->add_child(new FilterTree(new std::string("port " + std::to_string(port))));

        auto *portProtocolFilterTree = new FilterTree(FILTER_TREE_TYPE_OR);
        if ((has_tcp && has_udp) || (!has_tcp && !has_udp)) {
            portProtocolFilterTree->add_child(new FilterTree(new std::string("tcp")));
            portProtocolFilterTree->add_child(new FilterTree(new std::string("udp")));
        } else if (has_tcp) {
            portProtocolFilterTree->add_child(new FilterTree(new std::string("tcp")));
        } else if (has_udp) {
            portProtocolFilterTree->add_child(new FilterTree(new std::string("udp")));
        }

        portFilterTree->add_child(portProtocolFilterTree);

        filterTree->add_child(portFilterTree);
    } else {
        if (has_tcp) { filterTree->add_child(new FilterTree(new std::string("tcp"))); }

        if (has_udp) { filterTree->add_child(new FilterTree(new std::string("udp"))); }
    }

    bool has_icmp4 = arguments->get<bool>("--icmp4");

    if (has_icmp4) { filterTree->add_child(new FilterTree(new std::string("icmp"))); }

    bool has_icmp6 = arguments->get<bool>("--icmp6");

    if (has_icmp6) {
        auto *icmp6EchoReqResFilterTree = new FilterTree(FILTER_TREE_TYPE_AND);

        icmp6EchoReqResFilterTree->add_child(new FilterTree(new std::string("icmp6")));
        icmp6EchoReqResFilterTree->add_child(new FilterTree(FILTER_TREE_TYPE_OR));
        icmp6EchoReqResFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 128")));
        icmp6EchoReqResFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 129")));

        filterTree->add_child(icmp6EchoReqResFilterTree);
    }

    bool has_arp = arguments->get<bool>("--arp");

    if (has_arp) { filterTree->add_child(new FilterTree(new std::string("arp"))); }

    bool has_ndp = arguments->get<bool>("--ndp");

    if (has_ndp) {
        auto *ndpFilterTree = new FilterTree(FILTER_TREE_TYPE_AND);

        ndpFilterTree->add_child(new FilterTree(new std::string("icmp6")));
        ndpFilterTree->add_child(new FilterTree(FILTER_TREE_TYPE_OR));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 133")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 134")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 135")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 136")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 137")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 148")));
        ndpFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 149")));

        filterTree->add_child(ndpFilterTree);
    }

    bool has_igmp = arguments->get<bool>("--igmp");

    if (has_igmp) { filterTree->add_child(new FilterTree(new std::string("igmp"))); }

    bool has_mld = arguments->get<bool>("--mld");

    if (has_mld) {
        auto *mldFilterTree = new FilterTree(FILTER_TREE_TYPE_AND);

        mldFilterTree->add_child(new FilterTree(new std::string("icmp6")));
        mldFilterTree->add_child(new FilterTree(FILTER_TREE_TYPE_OR));
        mldFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 130")));
        mldFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 131")));
        mldFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 132")));
        mldFilterTree->get_children()->at(1)->add_child(new FilterTree(new std::string("icmp6[0] == 143")));

        filterTree->add_child(mldFilterTree);
    }

    return filterTree->generate_filter();
}

/**
 * Format timeval of the packet to RFC3339 timestamp
 * @param timestamp timeval of the packet
 * @return RFC3339 timestamp string
 */
std::string format_timestamp(const struct timeval &timestamp) {
    auto time = std::chrono::system_clock::from_time_t(timestamp.tv_sec);
    auto in_time_t = std::chrono::system_clock::to_time_t(time);

    std::stringstream milliseconds_ss;
    milliseconds_ss << std::setw(3) << std::setfill('0') << timestamp.tv_usec / 1000;

    std::stringstream timezone_ss;
    timezone_ss << std::put_time(std::localtime(&in_time_t), "%z");

    auto timezone_str = timezone_ss.str();
    timezone_str.insert(3, ":");

    std::stringstream ss;
    ss << std::put_time(std::localtime(&in_time_t), "%Y-%m-%dT%H:%M:%S") << "." << milliseconds_ss.str()
       << timezone_str;

    return ss.str();
}

/**
 * Format IP addresses of the packet
 * @param packet packet
 * @param ether_type ether type
 * @return pair of source and destination IP addresses
 */
std::pair<std::string, std::string> format_ip_addresses(u_char *packet, uint16_t ether_type) {
    if (ntohs(ether_type) == ETHERTYPE_IP) {
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];

        inet_ntop(AF_INET, packet + 26, src_ip, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, packet + 30, dst_ip, INET_ADDRSTRLEN);

        return std::make_pair(src_ip, dst_ip);
    } else if (ntohs(ether_type) == ETHERTYPE_IPV6) {
        auto *ip_header = (struct ip6_hdr *) (packet + sizeof(struct ether_header));
        char src_ip[INET6_ADDRSTRLEN];
        char dst_ip[INET6_ADDRSTRLEN];
        inet_ntop(AF_INET6, &ip_header->ip6_src, src_ip, INET6_ADDRSTRLEN);
        inet_ntop(AF_INET6, &ip_header->ip6_dst, dst_ip, INET6_ADDRSTRLEN);
        return std::make_pair(src_ip, dst_ip);
    } else if (ntohs(ether_type) == ETHERTYPE_ARP) {
        char src_ip[INET_ADDRSTRLEN];
        char dst_ip[INET_ADDRSTRLEN];

        inet_ntop(AF_INET, packet + 28, src_ip, INET_ADDRSTRLEN);
        inet_ntop(AF_INET, packet + 38, dst_ip, INET_ADDRSTRLEN);

        return std::make_pair(src_ip, dst_ip);
    }

    return std::make_pair("unknown", "unknown");
}

/**
 * Format MAC addresses of the packet
 * @param ether_shost ethernet source address
 * @param ether_dhost ethernet destination address
 * @return pair of source and destination MAC addresses
 */
std::pair<std::string, std::string> format_mac(uint8_t *ether_shost, uint8_t *ether_dhost) {
    std::stringstream src_mac_ss;
    std::stringstream dst_mac_ss;

    for (int i = 0; i < 6; i++) {
        src_mac_ss << std::hex << std::setw(2) << std::setfill('0') << (int) ether_shost[i];
        dst_mac_ss << std::hex << std::setw(2) << std::setfill('0') << (int) ether_dhost[i];

        if (i < 5) {
            src_mac_ss << ":";
            dst_mac_ss << ":";
        }
    }

    return std::make_pair(src_mac_ss.str(), dst_mac_ss.str());
}

/**
 * Format ports of the packet
 * @param packet packet
 * @param ether_type ether type
 * @return pair of source and destination ports
 */
std::pair<int, int> *format_ports(const u_char *packet, uint16_t ether_type) {
    std::pair<int, int> *ports = nullptr;

    uint16_t protocol = ntohs(ether_type);

    if (protocol == ETHERTYPE_IP) {
        if (packet[23] == IPPROTO_TCP || packet[23] == IPPROTO_UDP) {
            ports = new std::pair<int, int>(ntohs(*((uint16_t *) (packet + 34))), ntohs(*((uint16_t *) (packet + 36))));
        }
    } else if (protocol == ETHERTYPE_IPV6) {
        if (packet[20] == IPPROTO_TCP || packet[20] == IPPROTO_UDP) {
            ports = new std::pair<int, int>(ntohs(*((uint16_t *) (packet + 54))), ntohs(*((uint16_t *) (packet + 56))));
        }
    }

    return ports;
}

/**
 * Print payload of the packet
 * @param packet packet
 * @param size size of the packet
 */
void print_payload(const u_char *packet, uint32_t size) {
    uint32_t i = 0;

    while (i < size) {
        std::cout << "0x" << std::hex << std::setfill('0') << std::setw(4) << i << ": ";
        for (uint32_t j = 0; j < 16; j++) {
            if (i + j < size) {
                std::cout << std::hex << std::setfill('0') << std::setw(2) << (int) packet[i + j] << " ";
            } else {
                std::cout << "   ";
            }
        }
        std::cout << " ";
        for (uint32_t j = 0; j < 16; j++) {
            if (i + j < size) {
                if (isprint(packet[i + j])) {
                    std::cout << packet[i + j];
                } else {
                    std::cout << ".";
                }
            }
        }
        std::cout << std::endl;
        i += 16;
    }
}

/**
 * Handler for each packet
 * @param args arguments
 * @param header header of the packet
 * @param packet packet
 */
void packet_handler(u_char *args, const struct pcap_pkthdr *header, const u_char *packet) {
    auto *eth_header = (struct ether_header *) packet;

    auto ip_addresses = format_ip_addresses((u_char *) packet, eth_header->ether_type);
    auto mac_addresses = format_mac(eth_header->ether_shost, eth_header->ether_dhost);

    std::cout << "timestamp: " << format_timestamp(header->ts) << std::endl;
    std::cout << "src MAC: " << mac_addresses.first << std::endl;
    std::cout << "dst MAC: " << mac_addresses.second << std::endl;
    std::cout << "frame length: " << std::dec << header->len << " bytes" << std::endl;

    std::cout << "src IP: " << ip_addresses.first << std::endl;
    std::cout << "dst IP: " << ip_addresses.second << std::endl;

    auto ports = format_ports(packet, eth_header->ether_type);

    if (ports != nullptr) {
        std::cout << "src port: " << ports->first << std::endl;
        std::cout << "dst port: " << ports->second << std::endl;
    }

    std::cout << std::endl;

    print_payload(packet, header->len);

    std::cout << std::endl;
}

/**
 * Init pcap and start sniffing
 * @param arguments arguments of the sniffer
 */
void sniff(argparse::ArgumentParser *arguments) {
    auto interface = arguments->get<std::string>("interface");

    try {
        is_device_valid(interface);
    } catch (...) {
        std::cout << "Invalid interface: " << interface << std::endl;
        exit(1);
    }

    const char *device = interface.c_str();
    char errbuf[PCAP_ERRBUF_SIZE];
    bpf_u_int32 subnet_mask, ip;

    if (pcap_lookupnet(device, &ip, &subnet_mask, errbuf) == -1) {
        printf("Could not get information for device: %s\n", device);
        ip = 0;
        subnet_mask = 0;
    }

    pcap_t *handle = pcap_open_live(device, BUFSIZ, 1, 1000, errbuf);
    if (handle == nullptr) {
        std::cout << "pcap_open_live() failed: " << errbuf << std::endl;
        return;
    }

    struct bpf_program filter {};
    auto filter_exp = generate_filter(arguments);

    if (pcap_compile(handle, &filter, filter_exp.c_str(), 0, ip) == -1) {
        std::cout << "pcap_compile() failed: " << pcap_geterr(handle) << std::endl;
        return;
    }

    if (pcap_setfilter(handle, &filter) == -1) {
        std::cout << "pcap_setfilter() failed: " << pcap_geterr(handle) << std::endl;
        return;
    }

    int number_of_packets = arguments->get<int>("-n");

    pcap_loop(handle, number_of_packets, packet_handler, nullptr);

    pcap_close(handle);
}

int main(int argc, char *argv[]) {
    auto *arguments = process_arguments(argc, argv);

    if (argc == 1) {
        print_interfaces();
        return 0;
    }

    auto interface = arguments->get<std::string>("-i");

    bool has_any_options =
            arguments->get<int>("-p") != 0 || arguments->get<bool>("--tcp") || arguments->get<bool>("--udp") ||
            arguments->get<bool>("--icmp4") || arguments->get<bool>("--icmp6") || arguments->get<bool>("--arp") ||
            arguments->get<bool>("--ndp") || arguments->get<bool>("--igmp") || arguments->get<bool>("--mld");

    if (interface == "-" && !has_any_options) {
        print_interfaces();
        return 0;
    }

    sniff(arguments);

    return 0;
}
